### DISCUSSION TITLE
"Smarter" behavior for "NormalizePath"

### DISCUSSION THREAD

**Discussion by tp971** [actix/actix-web#3258] [Ideas]
Currently, `NormalizePath` only supports either always appending a trailing slash or always trimming it before matching. I have a very specific use case that would be easier to implement with the following two features:

1. Only append a trailing slash if the current url does not match any route, but appending a slash would make it match a route (e.g. something like `TrailingSlash::TryAppend`).
2. Add an option to `NormalizePath` to redirect over HTTP instead of just re-routing internally.

This is basically the default behavior for file indexing in web servers like Apache: if the user goes to `/asdf` and `asdf` is a folder, it redirects to `/asdf/`. I think having this in the `NormalizePath` middleware would be a nice quality of life improvement.

**Comment by robjtede** [comment]
Fundamentally, a "smarter" behavior like what you've described is at odds with a middleware called "normalize path", so at the very least this should be a different middleware.

Old-school web servers like Apache match up their path semantics with file system semantics because that's what they were doing - serving files via HTTP. But even in that landscape it seems frought to allow trailing slashes for files and missing slashes for directories. Maybe this is just my prescriptivism but if you mean to access a directory then there should be a slash.

Modern web ideologies often use the term "resource" which is distinct from file systems and is more easily expressable without worrying about trailing slashes or not. This is why `NormalizePath` exists, it allows to to pick a style, stick to it throughout your routing tree, and not have to force that view on your clients.

The `Files` service _does_ behave a little like what you described, for the expected reasons, so if you need that for actual file system access then it's there.

All this said, I _think_ it's _possible_ to create a middleware that does some dynamic routing based on what's been registered in the router, but since routers are dynamic and nest-able, it would likely have some pretty sharp edges that make it just as difficult to use.

**Comment by robjtede** [comment]
In case it helps your use case, I also want to point out that any method which accepts [`IntoPatterns` such as `web::resource()`](https://docs.rs/actix-web/latest/actix_web/web/fn.resource.html) is able to register multiple paths.

```rust
web::resource(["/foo", "/foo/"])
```

It's more boilerplate for sure, but it's a lesser known feature of the manual routing setups that might be useful.

Similarly, for macro routing, [`#[routes]`](https://docs.rs/actix-web/latest/actix_web/attr.routes.html) allows multiple path registration.

**Reply by tp971** [reply]
Okay, how about this: allow patterns like "/foo/?" (or something similar) that match both "/foo" and "/foo/", making it unnecessary to declare both patterns. Also, put a method somewhere to check whether the URL has a trailing slash or not (maybe in `actix_web::web::Path`?)

**Reply by tp971** [reply]
What I'm thinking of in my case is something like this:
```rs
#[get("/some/route/?", wrap = "RedirectTrailingSlash::new()")]
```
(or however the `wrap` attribute is used, I haven't found an example for it yet)

**Reply by robjtede** [reply]
Interesting idea actually. Would you mind opening a feature request issue for one or both of these things?
